nixos/modules/services/system/self-deploy.nix

```diff
@@ -0,0 +1,174 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.self-deploy;
+
+  workingDirectory = "/var/lib/self-deploy";
+  repositoryDirectory = "${workingDirectory}/repo";
+  outPath = "${workingDirectory}/system";
+
+  gitWithRepo = "git -C ${repositoryDirectory}";
+
+  renderNixArgs = args:
+    let
+      toArg = key: value:
+        if builtins.isString value
+        then " --argstr ${lib.escapeShellArg key} ${lib.escapeShellArg value}"
+        else " --arg ${lib.escapeShellArg key} ${lib.escapeShellArg (toString value)}";
+    in
+      lib.concatStrings (lib.mapAttrsToList toArg args);
+
+  isPathType = x: lib.strings.isCoercibleToString x && builtins.substring 0 1 (toString x) == "/";
+
+in {
+  options.services.self-deploy = {
+    enable = lib.mkEnableOption "self-deploy";
+
+    nixFile = lib.mkOption {
+      type = lib.types.path;
+
+      default = "/default.nix";
+
+      description = ''
+      Path to nix file in repository. Leading '/' refers to root of
+      git repository.
+      '';
+    };
+
+    nixAttribute = lib.mkOption {
+      type = lib.types.str;
+
+      description = ''
+      Attribute of `nixFile` that builds the current system.
+      '';
+    };
+
+    nixArgs = lib.mkOption {
+      type = lib.types.attrs;
+
+      default = {};
+
+      description = ''
+      Arguments to `nix-build` passed as `--argstr` or `--arg` depending on
+      the type.
+      '';
+    };
+
+    switchCommand = lib.mkOption {
+      type = lib.types.enum [ "boot" "switch" "dry-activate" "test" ];
+
+      default = "switch";
+
+      description = ''
+      The `switch-to-configuration` subcommand used.
+      '';
+    };
+
+    repository = lib.mkOption {
+      type = with lib.types; oneOf [ path str ];
+
+      description = ''
+      The repository to fetch from. Must be properly formatted for git.
+
+      If this value is set to a path (must begin with `/`) then it's
+      assumed that the repository is local and the resulting service
+      won't wait for the network to be up.
+
+      If the repository will be fetched over SSH, you must add an
+      entry to `programs.ssh.knownHosts` for the SSH host for the fetch
+      to be successful.
+      '';
+    };
+
+    sshKeyFile = lib.mkOption {
+      type = with lib.types; nullOr path;
+
+      default = null;
+
+      description = ''
+      Path to SSH private key used to fetch private repositories over
+      SSH.
+      '';
+    };
+
+    branch = lib.mkOption {
+      type = lib.types.str;
+
+      default = "master";
+
+      description = ''
+      Branch to track
+
+      You can also specify a revision or tag, too, if you want to pin this
+      machine to a specific commit.
+      '';
+    };
```

Comment:
This `branch` value is used in a `git fetch` command, but `git fetch` requires a ref, not a rev. So I wouldn't say you can specify a revision.

Reply:
@lambdadog I'm familiar with this issue because I ran into it when writing `manual-deploy`.

If you tweak the options to more closely match what I have with `manual-deploy`'s CLI options, where the user explicitly says whether they're specifying a rev or a ref, then you can handle both cases.

You can't `fetch` a specific `rev`, so you'd just fetch everything from the remote, and then `checkout` that `rev`.

Reply:
I honestly don't feel it's necessary to deeply support the use-case of fetching a specific rev, seeing as the use-case for `self-deploy` is continuous deployment and a single revision isn't going to change in a way that requires regular redeployment. I'll change the wording of the description, but unless you can think of a compelling usecase for it I don't see myself implementing an alternate path just for this.

Reply:
Sounds good 👍  I agree a specific rev isn't the point of this service.